File: packages/meal-plan-engine/src/per-meal-matcher.ts
import type { Meal, MealMatchRequest, ScoredMeal } from "@caloriehero/shared-types";
import { filterMeals } from "./filters.js";
import { calculateScore, calculateDeviation } from "./scoring.js";
import { DEFAULT_SCORING_WEIGHTS } from "./constants.js";

/**
 * Full per-meal matching pipeline:
 * 1. Filter meals by allergens, dietary preferences, and category
 * 2. Score each meal against the macro targets
 * 3. Sort by score descending
 * 4. Apply limit
 *
 * Returns a ranked list of ScoredMeal objects.
 */
export function matchMeals(meals: Meal[], request: MealMatchRequest): ScoredMeal[] {
  const { constraints, allergies, dietaryPreferences, category, limit } = request;
  const { targets, weights } = constraints;

  // Step 1: Filter
  const filtered = filterMeals(meals, {
    allergies,
    dietaryPreferences,
    category,
  });

  // Step 2: Score
  const effectiveWeights = weights ?? DEFAULT_SCORING_WEIGHTS;

  const scored: ScoredMeal[] = filtered.map((meal) => ({
    meal,
    score: calculateScore(meal.nutritionalInfo, targets, effectiveWeights),
    deviation: calculateDeviation(meal.nutritionalInfo, targets),
  }));

  // Step 3: Sort by score descending (deterministic: stable sort with id tiebreak)
  scored.sort((a, b) => {
    const diff = b.score - a.score;
    if (diff !== 0) return diff;
    // Tiebreak by meal id for determinism
    return a.meal.id.localeCompare(b.meal.id);
  });

  // Step 4: Apply limit
  return scored.slice(0, limit);
}

File: backend/app/engine/per_meal_matcher.py
"""Per-meal matching pipeline: filter -> score -> sort -> top N."""

from app.engine.constants import DEFAULT_SCORING_WEIGHTS
from app.engine.filters import filter_meals
from app.engine.scoring import calculate_deviation, calculate_score
from app.engine.types import Meal, MealMatchRequest, ScoredMeal


def match_meals(meals: list[Meal], request: MealMatchRequest) -> list[ScoredMeal]:
    """Full per-meal matching pipeline.

    1. Filter meals by allergens, dietary preferences, and category
    2. Score each meal against the macro targets
    3. Sort by score descending (deterministic with id tiebreak)
    4. Apply limit
    """
    effective_weights = request.weights or DEFAULT_SCORING_WEIGHTS

    # Step 1: Filter
    filtered = filter_meals(
        meals,
        allergies=request.allergies,
        dietary_preferences=request.dietary_preferences,
        category=request.category,
    )

    # Step 2: Score
    scored = [
        ScoredMeal(
            meal=meal,
            score=calculate_score(meal.nutritional_info, request.targets, effective_weights),
            deviation=calculate_deviation(meal.nutritional_info, request.targets),
        )
        for meal in filtered
    ]

    # Step 3: Sort by score descending, tiebreak by id
    scored.sort(key=lambda s: (-s.score, s.meal.id))

    # Step 4: Apply limit
    return scored[: request.limit]
